### ISSUE TITLE
Nan when training "DIFFORMER-a"?

### ISSUE THREAD

**Issue by silentsaber** [qitianwu/DIFFormer#3]
Hi，您的工作很不错，我很感兴趣。不过我在尝试迁移到我的任务上时发现kernel='sigmoid'的时候训练中容易出现nan。

我查看了关于full_attention_conv的代码，发现其中涉及到sigmoid的部分，在计算分母的时候，attention_normalizer可能包含0元素。这会导致difformer-a在计算attention的时候会出现nan。(相比之下，在simple中分母在line38有`attention_normalizer += torch.ones_like(attention_normalizer) * N`可以保证分母不会为零。)
```python
elif kernel == 'sigmoid':
        # numerator
        attention_num = torch.sigmoid(torch.einsum("nhm,lhm->nlh", qs, ks))  # [N, L, H]

        # denominator
        all_ones = torch.ones([ks.shape[0]]).to(ks.device)
        attention_normalizer = torch.einsum("nlh,l->nh", attention_num, all_ones)
        attention_normalizer = attention_normalizer.unsqueeze(1).repeat(1, ks.shape[0], 1)  # [N, L, H]

        # compute attention and attentive aggregated results
        attention = attention_num / attention_normalizer
        attn_output = torch.einsum("nlh,lhd->nhd", attention, vs)  # [N, H, D]
```

请问有什么好的方法避免出现nan这种情况呢？(我考虑过给分母加上一个eps小量，或者采用torch.nan_to_num()这两种方法。想问问是否有更好的方法)

**Comment by qitianwu** [comment]
Hi，因为我们没有在这篇工作的几个实验场景里遇到nan的问题，我感觉可能您这里是进入sigmoid之前的特征太小了，导致sigmoid之后因为有限精度的原因就变成0了。所以可能可以试试在sigmoid之前加一个clip操作（torch里面应该是torch.clamp)

**Comment by silentsaber** [comment]
> Hi，因为我们没有在这篇工作的几个实验场景里遇到nan的问题，我感觉可能您这里是进入sigmoid之前的特征太小了，导致sigmoid之后因为有限精度的原因就变成0了。所以可能可以试试在sigmoid之前加一个clip操作（torch里面应该是torch.clamp)

非常感谢您的解答！如您所说，确实有可能是因为输入特征的量级太小，导致sigmoid之后的输出为0，我会尝试去解决这一问题。

**Comment by demo4ai** [comment]
Can sigmoid be replaced with softmax?

**Comment by qitianwu** [comment]
Yeah of course, you can try using softmax if the quadratic complexity is not a concern
